app/grandchallenge/reader_studies/signals.py

```diff
@@ -0,0 +1,40 @@
+from django.db.models.signals import m2m_changed
+from django.dispatch import receiver
+from guardian.shortcuts import assign_perm, remove_perm
+
+from grandchallenge.cases.models import Image
+from grandchallenge.reader_studies.models import ReaderStudy
+
+
+@receiver(m2m_changed, sender=ReaderStudy.images.through)
+def update_image_permissions(instance, action, reverse, model, pk_set, **_):
+    """
+    Assign or remove view permissions to the readers group when images
+    are added or remove to/from the reader study images. Handles reverse
+    relations and clearing.
+    """
+    if action not in ["post_add", "post_remove", "pre_clear"]:
+        # nothing to do for the other actions
+        return
+
+    if reverse:
+        images = Image.objects.filter(pk=instance.pk)
+        try:
+            reader_studies = model.objects.filter(pk__in=pk_set)
+        except TypeError:
+            # When a clear method is used pk_set is none and a TypeError raised
```

Comment:
Why not do a `if pk_set is None` then? has `pk_set` not been retrieved yet and this is deferring the database interaction/optimizing the number of requests? Otherwise, branching with exceptions is quite naughty and can better be rewritten as an if-statement.

Reply:
Fixed. It didn't feel right to me (I don't really like all the work you need to do in this function, and that `pk_set is None` means you clear the entire through table, but I trust Django not to change this API and have it well tested). But, `pk_set` is already resolved, so doing this test saves actually saves a database trip in the `_clear` action case.